### PR TITLE
[typescript] fix: `nullable` enums should not serialize a `null` value to a string

### PR DESCRIPTION
**Pull request by joscha**
nullable enums with a value of `null` (e.g. `['a', 'b', null]`) currently receive an additional (incorrect) value `Null = 'null'`.

This issue was introduced in https://github.com/OpenAPITools/openapi-generator/pull/19027
Sample is here: https://github.com/OpenAPITools/openapi-generator/pull/19027#issuecomment-2333782021

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh ./bin/configs/*.yaml
  ./bin/utils/export_docs_generators.sh
  ``` 
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (upcoming 7.6.0 minor release - breaking changes with fallbacks), `8.0.x` (breaking changes without fallbacks)
- [x] If your PR is targeting a particular programming language, @mention the [technical committee] @TiFu (2017/07) @taxpon (2017/07) @sebastianhaas (2017/07) @kenisteward (2017/07) @Vrolijkx (2017/09) @macjohnny (2018/01) @topce (2018/10) @akehir (2019/07) @petejohansonxo (2019/11) @amakhrov (2020/02) @davidgamero (2022/03) @mkusaka (2022/04)

### CODE REVIEW

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java
Comment (joscha) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java:3950]: I believe this can be rolled back, as `isValueNull` is sufficient.
The current model before this change is:

```
      "_enum" : [ "affinity-data", "dealroom", null ],
      "allowableValues" : {
        "enumVars" : [ {
          "name" : "AffinityData",
          "isValueNull" : false,
          "isString" : false,
          "value" : "'affinity-data'"
        }, {
          "name" : "Dealroom",
          "isValueNull" : false,
          "isString" : false,
          "value" : "'dealroom'"
        }, {
          "name" : "Null",
          "isValueNull" : true,
          "isString" : false,
          "value" : "'null'"
        } ],
        "values" : [ "affinity-data", "dealroom", null ]
      },
```

you can see there is an oddity in `allowableValues`, where the value is already serialized to `'null'`, however, #19027 didn't change this I believe, as it only touched `modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/TypeScriptClientCodegen.java` not the `DefaultCodegen`.

The `'null'` value is set here: https://github.com/joscha/openapi-generator/blob/85763cdb083a0803aba9fcbe88161326565667e1/modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java#L6639

maybe the surrounding loop https://github.com/joscha/openapi-generator/blob/85763cdb083a0803aba9fcbe88161326565667e1/modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java#L6626 needs the guard as well and we just remove the value from the possible enum values altogether? WDYT?

File: modules/openapi-generator/src/test/java/org/openapitools/codegen/typescript/fetch/TypeScriptFetchModelTest.java
Comment (joscha) [inline modules/openapi-generator/src/test/java/org/openapitools/codegen/typescript/fetch/TypeScriptFetchModelTest.java:384]: this test seems to be identical to `modules/openapi-generator/src/test/java/org/openapitools/codegen/php/PhpModelTest.java` - it seems maybe these files / tests were duplicated at some point. Might be worth moving this up onto a higher level and remove the duplication?

File: modules/openapi-generator/src/test/resources/3_0/typescript/19027-regression.yaml
Comment (joscha) [inline modules/openapi-generator/src/test/resources/3_0/typescript/19027-regression.yaml:28]: this is the line of the fixture that shows the issue

File: samples/openapi3/client/petstore/typescript/builds/nullable-enum/models/Response.ts
Comment (joscha) [inline samples/openapi3/client/petstore/typescript/builds/nullable-enum/models/Response.ts:44]: Before the change in this PR, there is an additional line
```
Null = 'null'
```

File: samples/openapi3/client/petstore/typescript/builds/nullable-enum/models/Response.ts
Comment (joscha) [inline samples/openapi3/client/petstore/typescript/builds/nullable-enum/models/Response.ts:19]: nullability (e.g. the possible value `null`) is declared here already.

File: modules/openapi-generator/src/test/resources/3_0/typescript/19027-regression.yaml
Comment (joscha) [inline modules/openapi-generator/src/test/resources/3_0/typescript/19027-regression.yaml:17]: I wasn't sure whether it is acceptable to just remove it, but this part of the definition is not actually needed to show the issue.

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java
Comment (macjohnny) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java:3950]: @wing328 whats the core guidance here?

Comment (joscha) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java:3950]: I am just trying to guard the second part:

```diff
diff --git a/modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java b/modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java
index f27e4a400e4..9215daee503 100644
--- a/modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java
+++ b/modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java
@@ -6630,6 +6630,9 @@ public class DefaultCodegen implements CodegenConfig {
                 : 0;
 
         for (Object value : values) {
+            if(value == null) {
+                continue;
+            }
             Map<String, Object> enumVar = new HashMap<>();
             String enumName = truncateIdx == 0
                     ? String.valueOf(value)
```

and see how it will affect the samples. My guess is that it might be the better solution, as we don't introduce another template variable, keep the model more sane and it will fix all future generators as well.

Comment (macjohnny) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java:3950]: yes that would be nice, and it would work for all generators without modifying the templates.

Comment (joscha) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java:3950]: Okay, I will make the change later.

Just OOI: is there a better way to crate fixtures for a very specific part of the generation? It feels like quite the overkill to generate a whole scaffold with 10+ files just to assert one tiny part of it.
For the model change in this PR I could write a unit test, but generally, is there a way to create a partial scaffold in any way?

Comment (joscha) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java]: just pushed 01209d53266 with the changes.

File: samples/openapi3/client/petstore/typescript/builds/nullable-enum/models/Response.ts
Comment (joscha) [inline samples/openapi3/client/petstore/typescript/builds/nullable-enum/models/Response.ts:43]: Before the change in this PR, there is an additional line
```
Null = 'null'
```

**Review by macjohnny — APPROVED** [review]

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java
Comment (macjohnny) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java]: looks good, thanks!

Comment (macjohnny) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java]: > Just OOI: is there a better way to crate fixtures for a very specific part of the generation? It feels like quite the overkill to generate a whole scaffold with 10+ files just to assert one tiny part of it.
For the model change in this PR I could write a unit test, but generally, is there a way to create a partial scaffold in any way?

yeah unit tests are the more compact/preferred option for these specific aspects

Comment (joscha) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java]: > yeah unit tests are the more compact/preferred option for these specific aspects

I can change this PR to use unit tests if/when @wing328 approves and we merge it.

**Comment by joscha** [comment]
@macjohnny @wing328 is there any more changes needed for this one to land, please?

**Comment by macjohnny** [comment]
we just need approval from the core team

**Comment by joscha** [comment]
> we just need approval from the core team

@wing328 could I get your 👀 on this one, please?

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java
Comment (wing328) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java:3953]: thanks for the PR 👍 

shall we also set `property.isNullable` to `true` when null is found?

Comment (joscha) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java:3953]: yes, I think that would make sense actually.

Comment (joscha) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java:3953]: can we do it in a separate PR?

Comment (wing328) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java:3953]: sure

Comment (joscha) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java:3953]:
```suggestion
                    property.isNullable = true;
                    continue;
```
or here you go :)

**Comment by joscha** [comment]
🎉 all green @wing328

File: bin/configs/typescript-consolidated-nullable-enum.yaml
Comment (wing328) [inline bin/configs/typescript-consolidated-nullable-enum.yaml:2]: for new TS sample folder, do we need to add it to the Circle CI script to ensure the output compiles without issues?

https://github.com/OpenAPITools/openapi-generator/blob/master/CI/circle_parallel.sh#L76

Comment (joscha) [reply bin/configs/typescript-consolidated-nullable-enum.yaml:2]: I guess, yes, would it make sense to determine this: https://github.com/OpenAPITools/openapi-generator/blob/5732e2758fa3a2bd854fbd12016b49bd5f6ed6ba/CI/circle_parallel.sh#L75-L104

automatically?

Comment (joscha) [reply bin/configs/typescript-consolidated-nullable-enum.yaml:2]: added bbe6d34 for now

Comment (wing328) [reply bin/configs/typescript-consolidated-nullable-enum.yaml:2]: not against automating anything

maybe we should take this opportunity to move some tests (which don't require a local petstore server) to github workflows instead

cc @macjohnny

**Pull request merged**
